src/MessagePack/Internal/UnicodeUtility.cs

```diff
@@ -0,0 +1,232 @@
+namespace MessagePack.Internal
+{
+    using System;
+    using System.Runtime.CompilerServices;
+    /// <summary>
+    /// unicode manupilation utilities
+    /// </summary>
+    public static class UnicodeUtility
+    {
+        [ThreadStatic]
+        const char UnicodeInvalidChar = (char)0xfffd;
+        static char[] Utf8Buffer;
+        /// <summary>
+        /// convert utf8 bytearray to string instance
+        /// </summary>
+        /// <param name="data">UTF-8 byte array(without BOM)</param>
+        /// <param name="offset">start position in data</param>
+        /// <param name="count">number of byte length</param>
+        /// <remarks>when the data contains only under U+7F and has long length(around 4k-8k), this is slower than Encoding.UTF8.GetBytes(need more optimization)</remarks>
+        /// <returns>converted string instance</returns>
+        public static unsafe string GetStringFromUtf8(byte[] data, int offset, int count)
+        {
+            if (data.Length < offset + count)
+            {
+                throw new IndexOutOfRangeException("offset + count exceeds on byte array length");
+            }
+            else if (count == 0)
+            {
+                return string.Empty;
+            }
+            fixed (byte* databeginptr = &data[offset])
+            {
+                byte* dataptr = databeginptr;
+                var endptr = dataptr + count;
+                if (count < 256)
+                {
+                    var buf = stackalloc char[count];
+                    char* iterptr = buf;
+                    while (UpdateCharUnsafe(ref dataptr, ref endptr, ref iterptr)) ;
+                    return new string(buf, 0, (int)(iterptr - buf));
+                }
+                else
+                {
+                    Utf8Buffer = Utf8Buffer != null && Utf8Buffer.Length >= count ? Utf8Buffer : new char[count];
```

Comment:
I can't allow cache huge buffer.
If count is too large, allocate huge buffer forever...

Reply:
how about set the cache size limit(const value, like 1MB)?
If it is ok, I will change.

Reply:
MessagePack for C# uses 64K as the threshold.
By the way, the official UTF8Encoding uses FastAllocatedString?, it is internal so very cheated:)
Perhaps, it avoids unnecessary allocation(But it is paying cost of two-path(calc char count -> decode)).
If this limit can not be exceeded and if the performance improvement range is small,
advantage of replacing encoding is lost(in that case, I can not merge it).

Reply:
[here is my benchmark code and result.](https://gist.github.com/itn3000/48ed2361eb6929857947e0386565bf20)
as you say, heap allocation case(>256bytes) without cash is slower and allocating more memory than UTF8Encoding.
but stack allocation case, faster than UTF8Encoding, and same memory usage.